homeassistant/components/sensor/gtfs.py

```diff
@@ -48,73 +51,143 @@
     vol.Required(CONF_DATA): cv.string,
     vol.Optional(CONF_NAME): cv.string,
     vol.Optional(CONF_OFFSET, default=0): cv.time_period,
+    vol.Optional(CONF_TOMORROW, default=False): cv.boolean,
 })
 
 
-def get_next_departure(sched, start_station_id, end_station_id, offset):
+def get_next_departure(sched, start_station_id, end_station_id, offset,
+                       include_tomorrow=False) -> Optional[dict]:
     """Get the next departure for the given schedule."""
     origin_station = sched.stops_by_id(start_station_id)[0]
     destination_station = sched.stops_by_id(end_station_id)[0]
 
     now = datetime.datetime.now() + offset
-    day_name = now.strftime('%A').lower()
-    now_str = now.strftime('%H:%M:%S')
-    today = now.strftime('%Y-%m-%d')
+    now_date = now.strftime(DATE_FORMAT)
+    yesterday = now - datetime.timedelta(days=1)
+    yesterday_date = yesterday.strftime(DATE_FORMAT)
+    tomorrow = now + datetime.timedelta(days=1)
+    tomorrow_date = tomorrow.strftime(DATE_FORMAT)
 
     from sqlalchemy.sql import text
 
-    sql_query = text("""
-    SELECT trip.trip_id, trip.route_id,
-           time(origin_stop_time.arrival_time) AS origin_arrival_time,
-           time(origin_stop_time.departure_time) AS origin_depart_time,
-           origin_stop_time.drop_off_type AS origin_drop_off_type,
-           origin_stop_time.pickup_type AS origin_pickup_type,
-           origin_stop_time.shape_dist_traveled AS origin_shape_dist_traveled,
-           origin_stop_time.stop_headsign AS origin_stop_headsign,
-           origin_stop_time.stop_sequence AS origin_stop_sequence,
-           time(destination_stop_time.arrival_time) AS dest_arrival_time,
-           time(destination_stop_time.departure_time) AS dest_depart_time,
-           destination_stop_time.drop_off_type AS dest_drop_off_type,
-           destination_stop_time.pickup_type AS dest_pickup_type,
-           destination_stop_time.shape_dist_traveled AS dest_dist_traveled,
-           destination_stop_time.stop_headsign AS dest_stop_headsign,
-           destination_stop_time.stop_sequence AS dest_stop_sequence
-    FROM trips trip
-    INNER JOIN calendar calendar
-               ON trip.service_id = calendar.service_id
-    INNER JOIN stop_times origin_stop_time
-               ON trip.trip_id = origin_stop_time.trip_id
-    INNER JOIN stops start_station
-               ON origin_stop_time.stop_id = start_station.stop_id
-    INNER JOIN stop_times destination_stop_time
-               ON trip.trip_id = destination_stop_time.trip_id
-    INNER JOIN stops end_station
-               ON destination_stop_time.stop_id = end_station.stop_id
-    WHERE calendar.{day_name} = 1
-               AND origin_depart_time > time(:now_str)
```

Comment:
Performing a non-indexed time() comparison to time(origin_stop_time.departure_time) (alias origin_depart_time) is very expensive.

The updated query forgoes this time() comparison altogether, performing the last-stage time filtering in code instead.